Bellman Training Loop

alanlujan91 opened the pull request:
This pull request makes several important updates to the neural network training and evaluation code for block policy and value networks in `src/skagent/ann.py`, as well as updates to the test suite and configuration. The main changes include a refactor to the value function computation for robustness, standardization of training epochs and function signatures, simplification of test cases, and the addition of a new test scenario.

**Neural network logic and robustness improvements:**

* Refactored the `value_function` method in `BlockValueNet` to robustly handle missing variables and derived computations. The function now attempts to compute derived variables using the block's transition method, gracefully falls back to available variables if needed, and uses zero tensors when information set variables are missing. This improves compatibility with complex models and incomplete test inputs.

**Training function standardization and simplification:**

* Standardized training functions (`train_block_policy_nn`, `train_block_value_nn`, and `train_block_value_and_policy_nn`) to use a default of 250 epochs and unified loss function signatures. The joint training function now uses a single optimizer and loss function for both networks, simplifying the training process and enabling shared objectives. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L567-R600) [[2]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L590-R624) [[3]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L608-R637) [[4]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L642-R678) [[5]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L660-R726)
* Updated the decision function signatures to use default empty dictionaries for `shocks_t` and `parameters`, ensuring consistent and error-free invocation. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L319-R319) [[2]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L342-L343) [[3]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L403-R401)

**Test suite simplification and expansion:**

* Simplified test cases in `tests/test_ann.py` by removing explicit `epochs` arguments and width parameters, relying on the new standardized defaults. This reduces boilerplate and makes tests easier to maintain. [[1]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL54-R53) [[2]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL75-R74) [[3]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL106-R105) [[4]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL134-R132) [[5]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL158-R156) [[6]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL184-R182) [[7]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL210-R208) [[8]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL234-R232) [[9]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL258-R256) [[10]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL284-R282) [[11]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL308-R306) [[12]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL343-R341) [[13]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL381)
* Added a new test scenario (`case_11`) in `tests/conftest.py` for a simple consumer block with log utility and stochastic shocks, expanding coverage for model calibration and grid configuration.

**Code style and minor improvements:**

* Improved print statements in training functions to use f-strings for clarity and consistency. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L590-R624) [[2]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L633-R662) [[3]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L660-R726)
* Minor type hinting and import updates for better readability and maintainability. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L272-R272) [[2]](diffhunk://#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R2-R4) [[3]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL18-L23)

These changes collectively enhance the flexibility, reliability, and maintainability of the neural network training pipeline and its associated tests.

alanlujan91 committed:
Refactor test cases in test_ann.py and test_maliar.py for improved clarity and structure. Consolidate multiple test cases, enhance documentation, and ensure consistent use of test setup methods. Update assertions to reflect expected outcomes accurately.

alanlujan91 committed:
Implement unified training for policy and value networks using a single loss function in the new bellman_training_loop. Update train_block_value_and_policy_nn to support joint optimization and modify related tests to validate the new functionality.

alanlujan91 committed:
Enhance Bellman residual estimation in maliar.py by computing derived variables before value network evaluation. Update test cases to include new case_11 for comprehensive coverage of policy training with Bellman loss. Ensure deterministic behavior in tests for reproducibility.

alanlujan91 committed:
dos2unix for proper diff?

alanlujan91 committed:
Remove deprecated Bellman equation loss tests from test_ann.py for improved clarity and focus on relevant test cases.

alanlujan91 committed:
Update hidden layer width to 16 in BlockPolicyNet and BlockValueNet, adjust training epochs to 250 in training functions, and refactor test cases for consistency. Enhance decision function handling in maliar.py and improve error handling in Bellman residual calculations.

alanlujan91 committed:
Remove redundant joint training function existence tests from test_ann.py and update test_maliar.py to use actual state variables in Bellman equation loss calculations. Simplify assertions for clarity.

alanlujan91 committed:
Update hidden layer width to 32 in BlockPolicyNet and BlockValueNet, adjust training epochs in test cases, and enhance Bellman loss function to support unified training of policy and value networks. Refactor related tests for clarity and consistency.

alanlujan91 commented:
@sbenthall still have to work on this tomorrow morning, but what I have concluded is that maliar_training_loop is pretty explicitly set up for policy training only, so I have created a bellman_training_loop for simultaneous training of policy and value function, which is what the bellman loss needs

sbenthall commented on src/skagent/algos/maliar.py:
```diff
@@ -253,6 +253,49 @@ def simulate_forward(
+    return torch.cat(params) if params else torch.tensor([])
+
+
+def _compute_parameter_difference(params1, params2):
```
unclear why this is maliar method specific. utils.py, or ann.py? and why 'private' ? (_ preceding)

sbenthall commented on src/skagent/algos/maliar.py:
```diff
@@ -263,40 +306,56 @@ def maliar_training_loop(
+        Loss function for policy training with signature (decision_function, input_grid) -> loss.
+        Must be a lifetime reward loss function. Cannot use Bellman residual loss since
+        this function only trains policy networks and does not create/track value networks.
+        Created by get_estimated_discounted_lifetime_reward_loss().
```
I see why you're going this way. But I'm not thrilled about a design where there's two methods with redundant code except for one piece. It's harder to maintain two rather than one. What design can unify these?

alanlujan91 committed:
Enhance Bellman equation implementation in maliar.py by adding numerical stability for control variables and improving error handling for NaN rewards. Update loss function to fully implement MMW Definition 2.10, including product of residuals and derivative terms. Refactor related tests in test_ann.py and test_maliar.py to reflect changes and ensure proper validation of the new functionality.

sbenthall commented on src/skagent/algos/maliar.py:
```diff
@@ -516,46 +607,184 @@ def get_bellman_equation_loss(
+    block : DBlock
+        The model definition
+    loss_function : callable
+        A unified loss function with signature (value_function, decision_function, input_grid) -> loss.
```
The difference in the loops seems to be about:
- the signature of the loss function
- the initialization of multiple networks.

I wonder if, in principle, it's possible to combine both the policy and value networks into a single network. Is that silly?

I'm asking because I recall that the MMW '21 algorithm is generic to the AiO operator used.

sbenthall commented on src/skagent/ann.py:
```diff
@@ -498,26 +496,56 @@ def value_function(self, states_t, shocks_t={}, parameters={}):
+        vals = parameters | states_t | shocks_t
+
+        # Try to compute derived variables using block transition
+        # If variables are missing for complex models, fall back to available variables
```
I don't understand the context where these fallbacks are used.

Are they needed for the Bellman training loop?

Are there separate automated tests that test for the execution of these code blocks?

Maybe these can be moved to a separate PR?

sbenthall commented on src/skagent/ann.py:
```diff
@@ -657,51 +685,44 @@ def train_block_value_and_policy_nn(
+            return loss_function(value_function, df, inputs)
+
+        joint_loss = aggregate_net_loss(
+            inputs, block_policy_nn.get_decision_function(), adapted_loss_function
```
This seems to be hacking around what was intended to be a unified type signature for loss functions.

If that abstraction is being broken it means the design needs better thinking.

sbenthall commented on src/skagent/ann.py:
```diff
@@ -657,51 +685,44 @@ def train_block_value_and_policy_nn(
-        Loss function for value training (takes value_function, input_grid)
+    loss_function : callable
+        Unified loss function that takes (value_function, decision_function, input_grid)
+        and trains both networks simultaneously
```
Is the Euler residual loss going to be like this?

If this is tightly coupled to the Bellman loss function, the abstractions are strange.

sbenthall commented on tests/test_ann.py:
```diff
@@ -767,33 +629,79 @@ def test_value_function_perfect_foresight(self):
+        # 1. New bellman_training_loop() that mirrors maliar_training_loop signature
+        # 2. Uses unified Bellman loss to train both networks simultaneously
+        # 3. Follows existing patterns and conventions
+        # 4. Demonstrates the comprehensive joint training that was missing before
```
I'm having trouble following all the changes.

Is there a test that shows that the loops hit an actual target policy/value function (say, in an analytically known case)?

alanlujan91 committed:
Refactor Bellman loss to use full Euler residuals

Rewrote the derivative term in get_bellman_equation_loss to compute the full Euler equation residuals as defined in MMW Definition 2.10, replacing the previous gradient proxy approach. Improved documentation to clarify the role of Bellman and Euler residuals in the loss function.

alanlujan91 committed:
Add Euler residual estimation and testing framework

Introduced a new function to estimate the Euler residual in maliar.py, enhancing the model's capability to compute first-order conditions without relying on a value function. Updated the pyproject.toml to include pytest as a development dependency. Added comprehensive unit tests for Euler residuals and related loss functions in test_euler.py and test_maliar.py, ensuring robust validation of the new functionality and adherence to existing patterns in the codebase.

alanlujan91 commented:
This has become a much longer PR after realizing that I needed to implement the Euler method, as it is used in the Bellman method. After working on this for a significant amount of time, I'm still not where I would like to be, and running out of ideas as to how to proceed. 

Using euler residuals in MMW is easy if you know those residuals and can use them symbolically. Having skagent figure out the residuals from the Dblock is turning out to be a headache. 

AI summary below: 


This pull request introduces improvements and refactoring to the neural network training and evaluation code for dynamic block models, with a focus on simplifying interfaces, making loss functions more flexible, and improving robustness for edge cases. It also updates benchmark models for analytical correctness and adds a new test case for consumer models with shocks. The most significant changes are grouped below.

### Neural Network Training and Loss Functions

* Refactored training functions in `src/skagent/ann.py` to use unified loss functions and joint optimizers, increasing epochs to 250 by default and switching to f-string logging for better readability. This allows simultaneous training of value and policy networks with a single loss function and simplifies the API for training. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L567-R600) [[2]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L590-R624) [[3]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L608-R637) [[4]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L633-R662) [[5]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L642-R678) [[6]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L660-R726)
* Updated the `decision_function` and `get_decision_function` interfaces to use empty dictionaries as default arguments for `shocks_t` and `parameters`, improving consistency and reducing the need for explicit argument passing. [[1]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L319-R319) [[2]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L342-L343) [[3]](diffhunk://#diff-e46e5b6e50c99aa1223143aba046be4cb9571a1bc9559548e6ff00bf462ac4c8L403-R401)
* Improved the `value_function` method to robustly handle missing variables and fallback gracefully using zeros when necessary, ensuring the network can handle incomplete information sets during testing and training.

### Benchmark Model Corrections

* Corrected the D-3 benchmark model in `src/skagent/models/benchmarks.py` to use wealth (`w`) instead of market resources (`m`), and set income (`y`) to zero for analytical solution validity. This change affects both the model dynamics and the analytical policy function. [[1]](diffhunk://#diff-ebd3db93fb059e00f74030d1a8c0d509ebf280a987a1f1122e74448636d73e30L202-R202) [[2]](diffhunk://#diff-ebd3db93fb059e00f74030d1a8c0d509ebf280a987a1f1122e74448636d73e30L211-R213) [[3]](diffhunk://#diff-ebd3db93fb059e00f74030d1a8c0d509ebf280a987a1f1122e74448636d73e30L225-R225) [[4]](diffhunk://#diff-ebd3db93fb059e00f74030d1a8c0d509ebf280a987a1f1122e74448636d73e30L239-R240)

### Test Suite Enhancements

* Added a new consumer test case with shocks (`case_11`) to `tests/conftest.py`, expanding coverage for dynamic models involving stochastic elements.
* Simplified test calls in `tests/test_ann.py` by removing explicit width and epochs parameters from `BlockPolicyNet` and `train_block_policy_nn`, relying on new defaults for easier test maintenance. [[1]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL54-R53) [[2]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL75-R74) [[3]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL106-R105) [[4]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL134-R132) [[5]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL158-R156) [[6]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL184-R182) [[7]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL210-R208) [[8]](diffhunk://#diff-eb9fe7ab9a93c7126ea75284b0ee4bd18f9877eaa6839cb87b7bd0c58ae710caL234-R232)

### Dependency Management

* Added a `[dependency-groups]` section with a `dev` group in `pyproject.toml` for development dependencies, specifically including `pytest` for testing, which helps streamline environment setup for contributors.

sbenthall commented:
Thanks for working on this.

Let's break the problem down into incremental improvements and concrete
tests for each new function.

That way it will be easier to accept changes and then split the work into
parallel streams this coming week.

If it makes sense we can pair program on the hardest parts of the problem.

I think you've suggested using explicit residuals instead of autograd and
that's certainly better than nothing.

I also wonder if the envelope condition could be used instead of the euler
residual. Maybe that's worth testing.

On Sat, Aug 16, 2025, 7:27 PM Copilot ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> Pull Request Overview
>
> This pull request implements comprehensive Bellman residual minimization
> training for neural networks in the skagent library, enhancing robustness
> and standardizing training interfaces. The changes add value function
> approximation capabilities, standardize training epochs, and introduce
> joint training for policy and value networks using MMW Definition 2.10
> methodologies.
>
> Key changes include:
>
>    - Enhanced BlockValueNet with robust variable handling and fallback
>    mechanisms for incomplete model specifications
>    - Standardized all training functions to use 250 epochs by default
>    with unified loss function signatures
>    - Added complete Bellman training loop (bellman_training_loop) for
>    joint policy/value network optimization
>    - Simplified test suite by removing explicit parameter specifications
>    and relying on new defaults
>
> Reviewed Changes
>
> Copilot reviewed 10 out of 10 changed files in this pull request and
> generated 5 comments.
> Show a summary per file
> File Description
> src/skagent/ann.py Enhanced BlockValueNet robustness, standardized
> training epochs to 250, and unified joint training interface
> src/skagent/algos/maliar.py Added complete Bellman residual functions,
> Euler equation loss, and bellman_training_loop for joint optimization
> tests/test_ann.py Simplified test cases by removing explicit epochs/width
> parameters, added comprehensive Bellman training tests
> tests/conftest.py Added case_11 simple consumer model with Bellman grid
> configuration for enhanced test coverage
> tests/test_maliar.py Removed entire file - functionality moved to modular
> test files in tests/algos/ directory
> tests/algos/test_maliar.py New focused tests for maliar training loop and
> lifetime reward functions
> tests/algos/test_euler.py New comprehensive tests for Euler equation
> residual and loss functions
> tests/algos/test_bellman.py New extensive tests for Bellman residual,
> joint training, and value network functionality
> src/skagent/models/benchmarks.py Updated D-3 model variable names from
> 'm' to 'w' for consistency with analytical solution
> pyproject.toml Added dev dependency group with pytest for improved
> development workflow
> ------------------------------
>
> *Tip:* Customize your code reviews with copilot-instructions.md. Create
> the file
> <http:///scikit-agent/scikit-agent/new/main/.github?filename=copilot-instructions.md>
> or learn how to get started
> <https://docs.github.com/en/copilot/customizing-copilot/adding-repository-custom-instructions-for-github-copilot>
> .
> ------------------------------
>
> In src/skagent/ann.py
> <https://github.com/scikit-agent/scikit-agent/pull/104#discussion_r2280643853>
> :
>
> > @@ -316,7 +316,7 @@ def _setup_bound(self, bound_func, bound_name):
>              return vec_func, param_to_column
>          return None, None
>
> -    def decision_function(self, states_t, shocks_t, parameters):
> +    def decision_function(self, states_t, shocks_t={}, parameters={}):
>
> Using mutable default arguments (empty dictionaries) can lead to
> unexpected behavior. Consider using None as default and creating new
> dictionaries inside the function.
> ⬇️ Suggested change
>
> -    def decision_function(self, states_t, shocks_t={}, parameters={}):
> +    def decision_function(self, states_t, shocks_t=None, parameters=None):
>
> ------------------------------
>
> In src/skagent/ann.py
> <https://github.com/scikit-agent/scikit-agent/pull/104#discussion_r2280643854>
> :
>
> > @@ -400,7 +398,7 @@ def forward(self, x):
>          return x2
>
>      def get_decision_function(self):
> -        def df(states_t, shocks_t, parameters):
> +        def df(states_t, shocks_t={}, parameters={}):
>
> Using mutable default arguments (empty dictionaries) can lead to
> unexpected behavior. Consider using None as default and creating new
> dictionaries inside the function.
> ⬇️ Suggested change
>
> -        def df(states_t, shocks_t={}, parameters={}):
> +        def df(states_t, shocks_t=None, parameters=None):
> +            if shocks_t is None:
> +                shocks_t = {}
> +            if parameters is None:
> +                parameters = {}
>
> ------------------------------
>
> In src/skagent/algos/maliar.py
> <https://github.com/scikit-agent/scikit-agent/pull/104#discussion_r2280643857>
> :
>
> > +            next_states_grad[sym] = torch.tensor(val, requires_grad=True)
> +
> +    # Get next-period controls for reward computation
> +    next_controls = df(next_states_grad, shocks_t_plus_1, parameters)
> +
> +    # Add numerical stability for next-period controls
> +    next_controls_safe = {}
> +    for sym, val in next_controls.items():
> +        if isinstance(val, torch.Tensor):
> +            next_controls_safe[sym] = torch.clamp(val, min=1e-8)
> +        else:
> +            next_controls_safe[sym] = val
> +
> +    # Compute next-period reward
> +    next_reward = rf(next_states_grad, shocks_t_plus_1, next_controls_safe, parameters)
> +    sum(next_reward[rsym] for rsym in reward_syms)
>
> This line computes the sum but doesn't assign it to a variable or use it.
> This appears to be unused code that should either be removed or assigned to
> a variable.
> ⬇️ Suggested change
>
> -    sum(next_reward[rsym] for rsym in reward_syms)
>
> ------------------------------
>
> In tests/conftest.py
> <https://github.com/scikit-agent/scikit-agent/pull/104#discussion_r2280643859>
> :
>
> > @@ -297,3 +299,43 @@
>          }
>      ),
>  }
> +
> +case_11 = {
> +    "block": DBlock(
> +        **{
> +            "name": "simple consumer",
> +            "shocks": {
> +                "theta": (Normal, {"mu": 0.0, "sigma": 1}),
> +            },
> +            "dynamics": {
> +                "m": lambda a, r, theta: a * r + math.e**theta,
>
> Using math.e**theta instead of torch.exp(theta) may cause type
> inconsistencies when theta is a tensor. Consider using torch.exp for
> consistency with PyTorch operations.
> ⬇️ Suggested change
>
> -                "m": lambda a, r, theta: a * r + math.e**theta,
> +                "m": lambda a, r, theta: a * r + torch.exp(theta),
>
> ------------------------------
>
> In tests/algos/test_bellman.py
> <https://github.com/scikit-agent/scikit-agent/pull/104#discussion_r2280643861>
> :
>
> > +from conftest import case_0, case_11
> +import conftest
> +import numpy as np
> +import os
> +import skagent.algos.maliar as maliar
> +import skagent.ann as ann
> +import skagent.grid as grid
> +import torch
> +import unittest
> +from skagent.ann import BlockValueNet
> +from skagent.algos.maliar import (
> +    get_bellman_equation_loss,
> +)
> +from skagent.models.benchmarks import d3_block, d3_calibration, d3_analytical_policy
> +
> +from algos.test_maliar import create_simple_linear_value_network
>
> Importing from a test module in another test module creates tight coupling
> and may cause import issues. Consider moving shared test utilities to a
> common test utilities module.
> ⬇️ Suggested change
>
> -from algos.test_maliar import create_simple_linear_value_network
> +from utils.test_utils import create_simple_linear_value_network
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-agent/scikit-agent/pull/104#pullrequestreview-3126104236>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQZEHTKRUAIFDDLKVCZSL3N6457AVCNFSM6AAAAACD3DZXQWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMRWGEYDIMRTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sbenthall commented:
I think this PR has been superceded by later work.
